### DISCUSSION TITLE
Data silently truncated when using psycopg3 (not with psycopg2)

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/sqlalchemy#9837] [Usage Questions] [answered]
Hi.

I'm not sure whether this is a bug or a feature.

When inserting long data in varchar(N) using psycopg3, the data is silently truncated to fit the varchar limit. With psycopg2, an exception would be raised.

After a bit of investigation, I found out it happens in SQLAlchemy, not psycopg:

```
# psycopg2
INSERT INTO test (num, data) VALUES (%(num)s, %(data)s) RETURNING test.id {'num': 100, 'data': 'abcdefghij'}

# psycopg3
INSERT INTO test (num, data) VALUES (%(num)s::INTEGER, %(data)s::VARCHAR(5)) RETURNING test.id {'num': 100, 'data': 'abcdefghij'}
```

Apparently, some type cast/coercion was introduced. I find it a bit strange to silently truncate input data.

Is this intended?

-------------------------------------------------------------------------------------

Here's a full example:

```py
import psycopg
import sqlalchemy as sqla
from sqlalchemy import orm


# Pure psycopg3: create table and check that an error is returned when inserting data that is too long
with psycopg.connect("dbname=test-psycopg3 user=jerome password=xxx") as conn:

    with conn.cursor() as cur:

        cur.execute("""
            CREATE TABLE test (
                id serial PRIMARY KEY,
                num integer,
                data varchar(5))
            """)

        conn.commit()

        # Raises psycopg.errors.StringDataRightTruncation
        try:
            cur.execute(
                "INSERT INTO test (num, data) VALUES (%s, %s)", (100, "abcdefghij")
            )
        except psycopg.errors.StringDataRightTruncation as exc:
            print(exc)


class Base(orm.DeclarativeBase):
    pass


class Test(Base):
    __tablename__ = "test"

    id = sqla.Column(sqla.Integer, primary_key=True)
    num = sqla.Column(sqla.Integer())
    data = sqla.Column(sqla.String(5))


# SQLA + psycopg3
engine = sqla.create_engine("postgresql+psycopg://jerome:xxx@localhost:5432/test-psycopg3")
with orm.Session(engine) as session:

    # INSERT INTO test (num, data) VALUES (%(num)s::INTEGER, %(data)s::VARCHAR(5)) RETURNING test.id
    # {'num': 100, 'data': 'abcdefghij'}
    test = Test(num=100, data="abcdefghij")
    session.add(test)
    session.commit()

    # Data is silently truncated
    test = session.query(Test).first()
    assert test.data == "abcdefghij"[:5]


# SQLA + psycopg2
engine = sqla.create_engine("postgresql+psycopg2://jerome:xxx@localhost:5432/test-psycopg3")
with orm.Session(engine) as session:

    # INSERT INTO test (num, data) VALUES (%(num)s, %(data)s) RETURNING test.id
    # {'num': 100, 'data': 'abcdefghij'}
    test = Test(num=100, data="abcdefghij")
    session.add(test)
    # Raises sqlalchemy.exc.DataError: (psycopg2.errors.StringDataRightTruncation)
    try:
        session.commit()
    except sqla.exc.DataError as exc:
        print(exc)
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

For psycopg3 sqlalchemy adds cast to almost every type since it uses server side binding and postgresql is very picky regarding types, requiring types on most of them.

IIRC there was a similar issue that was fixed regarding this on the like side. What version of sqlalchemy are you using? cc @zzzeek

**Reply by lafrech** [reply]
Oh boy, you're right. I did update psycopg3 but my SQLAlchemy was a bit outdated.

I just bisected and the issue I'm showing above stops appearing in 2.0.8.

Which lead me to the changelog entry:

https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-9cfbfaeb532aa9eee370ade624279601

> Fixed critical regression in PostgreSQL dialects such as asyncpg which rely upon explicit casts in SQL in order for datatypes to be passed to the driver correctly, where a [String](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.String) datatype would be cast along with the exact column length being compared, leading to implicit truncation when comparing a VARCHAR of a smaller length to a string of greater length regardless of operator in use (e.g. LIKE, MATCH, etc.). The PostgreSQL dialect now omits the length from VARCHAR when rendering these casts.

https://github.com/sqlalchemy/sqlalchemy/issues/9511

Sorry about the noise.

**Reply by CaselIT** [reply]
great that is works also for inserts.
